### DISCUSSION TITLE
Question: feedback loop for temperature control via peltier

### DISCUSSION THREAD

**Discussion by kratsg** [python-control/python-control#562] [General]
This looks like a very nice (and well-developed) package. I haven't done the control systems through matlab (which a lot seems to be referenced) but I'd like to get started, maybe using this package to describe a system that provides feedback on the control of a peltier element (via modifying the voltage on a power supply).

I'm not entirely sure where to get started, but hoping I could get some tips or be pointed along the right path here. Nominally, the functionality would be that the current temperature, reference (goal) temperature, and current voltage as input -- with output being the new voltage to set to. [Perhaps an additional input would be a time-delta between measurements]. Thanks!

**Comment by bnavigator** [comment]
Hi,

of course there is the [package documentation](https://python-control.readthedocs.io/en/0.8.4/).
You can also start by comparing the examples given in [FBS](https://fbswiki.org/wiki/index.php/Feedback_Systems:_An_Introduction_for_Scientists_and_Engineers) with those [implemented for python-control](https://github.com/python-control/python-control/tree/master/examples), like the basic cruise control.

**Comment by kratsg** [comment]
> the basic cruise control.

Sorry, I did see the cruise control example, but that doesn't seem basic? Is this package not meant to be used for something like a simple PID controller? Because it would be really helpful to show how to use it for a simple application like that.

Personally, that's a very long example to try and digest, and I'm not even sure I can clearly formulate a set of equations in my case that describe the temperature flow of a complicated system.

**Reply by bnavigator** [reply]
> Is this package not meant to be used for something like a simple PID controller?

It is not fully clear, what you mean by "describe a system". This package is targeted to system control engineers (or students aspiring to become one) to aid in describing systems and design the controller, as well analyze the dynamical properties. You can do that with a PID controller (chapter 11 of FBS).

If you just want to run a loop executing the PID control, you don't need this package. Just write down the code in plain python, optionally collecting the data in numpy arrays and maybe with some help from `scipy.signal` functions.

> and I'm not even sure I can clearly formulate a set of equations

That makes it difficult to program anything, doesn't it? You will always need an adequate model of your system.

**Reply by kratsg** [reply]
> If you just want to run a loop executing the PID control, you don't need this package. Just write down the code in plain python, optionally collecting the data in numpy arrays and maybe with some help from `scipy.signal` functions.

Ok, so to clarify, this package is meant for more complex scenarios / systems? I've got an influxDB and all the time series data I need in order to be able to design feedback control. I was imagining that I could use a well-developed package like this to compose the PID controller and just specify the Kp, Ki, Kd and run it [but it sounds like that's not the goal for this package - so I misunderstood].

> That makes it difficult to program anything, doesn't it? You will always need an adequate model of your system.

This is the part I'm confused about. The main reason I suspect I wouldn't be able to write down an equation is that in the system I have, as the goal is to maintain some temperature with a peltier -- the system will heat up in a non-deterministic way based on the calculations being done on the processors. (But then again, I'm still new to this).

One could, for example, force 100% CPU usage to determine how fast the temperature goes up (step function) and then force 50% (step function), etc... and then plot this and model based on that. But that doesn't really help with getting the "equation of motion" that seems to be needed for this package.

**Reply by bnavigator** [reply]
> I was imagining that I could use a well-developed package like this to compose the PID controller and just specify the Kp, Ki, Kd 
[... ] One could, for example, force 100% CPU usage to determine how fast the temperature goes up (step function) and then force 50% (step function), etc... and then plot this ...

That's exactly the use case this package is designed for. But you first need to model and then you can analyze your model. You then adjust your model and see how it changes the results.

> and run it

Do you mean to simulate or do you want to drive an actual controller with input signals from sensors and output going to actuators?

> non-deterministic way based on the calculations

That sounds like a contradiction in itself.

In any case, whether you model your system with python-control or use any other software, you need a mathematical model of your system. A standard PID controller is only one part of the system. The transfer function for a standard PID employing the Kp, Ki, Kd parameters can be found in any control text book, or even on the [Wikipedia page for "PID controller"](https://en.wikipedia.org/wiki/PID_controller#Laplace_form). But you still need to model what we call "the plant". How does your peltier element react to input of some sort (voltage) and how does it interact with the rest of the environment?

**Reply by kratsg** [reply]
I could really be misunderstanding still (sorry!)

> Do you mean to simulate or do you want to drive an actual controller with input signals from sensors and output going to actuators?

Drive an actual peltier (in this particular case). I don't even think I can build a model for this, because I'm not trying to drive a single car, but many different cars (to use the metaphor of the `cruise` example). In this case, we're driving the peltier (no actuators) by setting the voltage.

> But you still need to model what we call "the plant". How does your peltier element react to input of some sort (voltage) and how does it interact with the rest of the environment?

This is the part I'm still struggling with (and I'm not sure I want to try and model the incredibly non-linear portion of temperature propagating through various materials of different conductivities - unless that's what you're proposing?).

To clarify (including the "non-deterministic" part) -- we're cooling down a sensor-on-chip and the chip itself is performing scans to read out the sensor. Depending on the radioactive source in play, as well as the sensor performance and behavior, will heat up the entire chip. The goal of the PID controller would be to receive, as input, temperature of the system, and set the voltage of the peltier to some value in a range of [0, 5V], in order to achieve a setpoint temperature.

There's two things that (in my mind) make this different from the `cruise` example:
1. delayed response (the conductivity of these materials and how fast the temperature propagates is a function of the temperature of the system) where the peltier has to be driven a bit longer before seeing an impact on the temperature
2. thousands of chips+sensors will be placed in the housing [all with, of course, different behaviors which means different models] but the same software controller should still be able to cool down the chips to the setpoint temperature.

If you do say that this is possible, it would help if you could point me at something like FBS talking specifically about driving/controlling anything temperature-related. I can find many examples of mechanical systems involving speed/thrust/motion -- but I'm at a loss for anything temperature-related. If I was going to build a model, I'd need to do Finite-Element-Analysis for the thermal portions here in order to model it properly... Or am I somehow overcomplicating things?

**Reply by bnavigator** [reply]
> In this case, we're driving the peltier (no actuators) by setting the voltage.

So the peltier *is* the actuator.

But I think I also misunderstood you or still do. Now it sounds like you are more looking for software to do realtime calculations including talking to the interfaces where you get the sensor inputs and set the peltier voltage output. Neither python-control nor MATLAB, which you mentioned are designed for that. (Although with some tweaks, Python or MATLAB may be abused for it)

To simulate anything, you need to model. The degree of approximation used for the model is at the discretion of the engineer. So is the choice of type of controller. It does not have to be a PID. A Bang-Bang controller would be even simpler to implement (but not to model). On the other hand, if robustness against different chip behaviors is key, you might even go overboard with the topic of "robust control".

**Reply by kratsg** [reply]
> But I think I also misunderstood you or still do. Now it sounds like you are more looking for software to do realtime calculations including talking to the interfaces where you get the sensor inputs and set the peltier voltage output. Neither python-control nor MATLAB, which you mentioned are designed for that. (Although with some tweaks, Python or MATLAB may be abused for it)

I think this is exactly what I want. Ok, I wasn't aware. I thought one _could_ use `python-control` to actually drive something, but if it's purely a simulation-based tool [e.g. to understand exactly your system that you've modeled] then it's different than what I was looking for!

I think if I was to use the `cruise` metaphor (poorly), this would be like someone giving you a parking lot full of cars of various makes/models and various states of wear and tear -- and giving you the instruction: "give them all cruise control functionality". This probably answers my question then (sorry for misrepresenting it, as I'm not super familiar with the industry-specific jargon here).

**Reply by josujugo** [reply]
Hi, 

Maybe you should see this one:

https://github.com/robertobucher/pysimCoder

Best regards

josu

**Reply by kratsg** [reply]
Thanks Josu! I will look into that. I found this one as well (https://github.com/m-lundberg/simple-pid/) so there seems to be examples I can look at.
